packages/components/src/number-control/test/index.tsx

```diff
@@ -0,0 +1,616 @@
+			} );
+
+			render( <StatefulNumberControl value={ 877.5309 } step="any" /> );
+
+			const input = getInput();
+			await user.click( input );
+			await user.keyboard( '{Shift>}[ArrowDown]{/Shift}' );
+
+			expect( input ).toHaveValue( 867.5309 );
+		} );
+
+		it( 'should decrement by custom shiftStep on key DOWN + shift press', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+
+			render( <StatefulNumberControl value={ 5 } shiftStep={ 100 } /> );
+
+			const input = getInput();
+			await user.click( input );
+			await user.keyboard( '{Shift>}[ArrowDown]{/Shift}' );
+
+			expect( input ).toHaveValue( -100 );
+		} );
+
+		it( 'should decrement but be limited by min on shiftStep', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+
+			render(
+				<StatefulNumberControl
+					value={ 5 }
+					shiftStep={ 100 }
+					min={ 4 }
+				/>
+			);
+
+			const input = getInput();
+			await user.click( input );
+			await user.keyboard( '{Shift>}[ArrowDown]{/Shift}' );
+
+			expect( input ).toHaveValue( 4 );
+		} );
+
+		it( 'should not decrement by shiftStep if disabled', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+
+			render(
+				<StatefulNumberControl
+					value={ 5 }
+					shiftStep={ 100 }
+					isShiftStepEnabled={ false }
+				/>
+			);
+
+			const input = getInput();
+			await user.click( input );
+			await user.keyboard( '{Shift>}[ArrowDown]{/Shift}' );
+
+			expect( input ).toHaveValue( 4 );
+		} );
+	} );
+
+	describe( 'custom spin buttons', () => {
+		test.each( [
+			undefined,
+			'none',
+			'native',
+		] as NumberControlProps[ 'spinControls' ][] )(
+			'should not appear when spinControls = %s',
+			( spinControls ) => {
+				render( <NumberControl spinControls={ spinControls } /> );
+				expect(
+					screen.queryByLabelText( 'Increment' )
+				).not.toBeInTheDocument();
+				expect(
+					screen.queryByLabelText( 'Decrement' )
+				).not.toBeInTheDocument();
+			}
+		);
+
+		test.each( [
+			[ 'up', '1', {} ],
+			[ 'up', '2', { value: '1' } ],
+			[ 'up', '12', { value: '10', step: '2' } ],
+			[ 'up', '10', { value: '10', max: 10 } ],
+			[ 'down', '-1', {} ],
+			[ 'down', '1', { value: '2' } ],
+			[ 'down', '10', { value: '12', step: '2' } ],
+			[ 'down', '10', { value: '10', min: 10 } ],
+		] )(
+			'should spin %s to %s when props = %o',
+			async ( direction, expectedValue, props ) => {
+				const user = userEvent.setup( {
+					advanceTimers: jest.advanceTimersByTime,
+				} );
+				const onChange = jest.fn();
+				render(
+					<NumberControl
+						{ ...props }
+						spinControls="custom"
+						onChange={ onChange }
+					/>
+				);
+				await user.click(
+					screen.getByLabelText(
+						direction === 'up' ? 'Increment' : 'Decrement'
+					)
+				);
+				expect( onChange ).toHaveBeenCalledWith( expectedValue, {
+					event: expect.anything(),
```

Comment:
Could we make this more precise? Something like this for example (untested):

```suggestion
					event: expect.any( Event ),
```

Reply:
<details>

<summary>Unfortunately that doesn't seem to work</summary>

```
  ● NumberControl › custom spin buttons › should spin down to 1 when props = { value: '2' }

    expect(jest.fn()).toHaveBeenCalledWith(...expected)

    - Expected
    + Received

      "1",
      Object {
    -   "event": Any<Event>,
    +   "event": Object {
    +     "_reactName": "onClick",
    +     "_targetInst": null,
    +     "altKey": false,
    +     "bubbles": true,
    +     "button": 0,
    +     "buttons": 0,
    +     "cancelable": true,
    +     "clientX": 0,
    +     "clientY": 0,
    +     "ctrlKey": false,
    +     "currentTarget": <button
    +       aria-hidden="true"
    +       aria-label="Decrement"
    +       class="components-button css-1n3g25v-SpinButton ep09it40 is-small has-icon"
    +       tabindex="-1"
    +       type="button"
    +     >
    +       <svg
    +         aria-hidden="true"
    +         focusable="false"
    +         height="24"
    +         viewBox="0 0 24 24"
    +         width="24"
    +         xmlns="http://www.w3.org/2000/svg"
    +       >
    +         <path
    +           d="M7 11.5h10V13H7z"
    +         />
    +       </svg>
    +     </button>,
    +     "defaultPrevented": true,
    +     "detail": 1,
    +     "eventPhase": 3,
    +     "getModifierState": [Function modifierStateGetter],
    +     "isDefaultPrevented": [Function functionThatReturnsTrue],
    +     "isPropagationStopped": [Function functionThatReturnsFalse],
    +     "isTrusted": false,
    +     "metaKey": false,
    +     "movementX": 0,
    +     "movementY": 0,
    +     "nativeEvent": MouseEvent {
    +       "isTrusted": false,
    +     },
    +     "pageX": 0,
    +     "pageY": 0,
    +     "relatedTarget": null,
    +     "screenX": 0,
    +     "screenY": 0,
    +     "shiftKey": false,
    +     "target": <input
    +       autocomplete="off"
    +       class="components-input-control__input css-1hsyo1w-Input-dragStyles-fontSizeStyles-sizeStyles-customPaddings em5sgkm5"
    +       id="inspector-input-control-38"
    +       inputmode="numeric"
    +       max="Infinity"
    +       min="-Infinity"
    +       step="1"
    +       type="number"
    +       value="2"
    +     />,
    +     "timeStamp": 1669279340945,
    +     "type": "click",
    +     "view": null,
    +   },
      },

    Number of calls: 1

      606 |                                     )
      607 |                             );
    > 608 |                             expect( onChange ).toHaveBeenCalledWith( expectedValue, {
          |                                                ^
      609 |                                     event: expect.any( Event ),
      610 |                             } );
      611 |                     }

      at packages/components/src/number-control/test/index.tsx:608:24
```

</details>

Reply:
Looks like it could work with (untested):

```suggestion
					event: expect.any( Object ),
```

Not perfect, but a bit more specific. 

Or we could even (untested):

```suggestion
					event: expect.objectContaining( {
						target: expect.any(),
						type: expect.any( String )
					} ),
```

Reply:
All those are just an optional suggestion and totally non-blocking FWIW.

Reply:
Good point! Done in https://github.com/WordPress/gutenberg/pull/45990/commits/1bbb118fcaaa5a541d8edb6cf30935a8832f56bf